documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -111,8 +131,49 @@ Threading rules for the reference implementation (if any) will be explicitly doc
 
 **INFORMATIVE TEXT:** For example, it is illegal for GLFW functions to be called anywhere except the thread `glfwInit` was called on, and it is illegal on some operating systems (such as macOS) for `glfwInit` to be called anywhere except the thread that called `main`.
 
+## Actors
+
+Some users will want to receive events in the order that they happen. Observing state does not allow us to do this if, for instance, a button was pressed and released between update calls - the state would have a net-zero change and thus we wouldn't be able to detect such events.
+
+This is solved by the actor model, which is implemented by the custom list types - the actor methods invoking the matching events.
+
+**INFORMATIVE TEXT:** A possible question is "why don't we just expose the events directly on the devices? Why do we need this actor model?". The answer to that is having events on the devices has considerations for the lifetime of the device objects, which the Silk.NET team wishes to leave unrestricted. For instance, when a device connects the input context will need to bind the events on each device itself and when a device disconnects the behaviour thereafter would be undefined because we have not explicitly allowed or disallowed the reuse of device objects, and we do not wish to. If a device object were to be reused when the device reconnects, those events would still be there. It's better to encapsulate these events in their own object using this actor model.
+
+```cs
+public interface IActor
+{
+    void HandleDeviceConnectionChanged(ConnectionEvent<IInputDevice> @event);
+}
+```
+
+The base `IActor` is simple, handling only device connections and device disconnections.
+
+`HandleDeviceConnectionChanged` must be called with a device and a value of `true` if the `ConnectionEvent.Device` has just been added to the `IInputBackend.Devices` list.
+
+`HandleDeviceConnectionChanged` must be called with a device and a value of `false` if the `ConnectionEvent.Device` has just been removed from the `IInputBackend.Devices` list.
+
+All actor methods are called in the order that the state changes happened in the underlying backend. For example, a "double click" would be mouse down, mouse up, mouse down, and mouse up again. Even if those events happened entirely between two update calls (thus resulting in a net-zero change of state, given the button was released at the last update and is still released at this update), the input backend is expected to queue up the events so they may be delivered to the actor in the order in which they occurred when `Update` is called **where possible**.
```

Comment:
clarification: *ordering* is optional but preferred, but consistency (raising the events at all) is required.

Reply:
Should ordering be optional on things like mouse button ups/downs though? Or should ordering be straight up required as to properly catch things like quickly shift-clicking?

Reply:
Yeah straight up required where the backend is able to provide it (will implement Kai's clarification)